sql/src/planner.rs

```diff
@@ -322,29 +343,133 @@ impl<'a, P: MetaProvider> PlannerDelegate<'a, P> {
+                ..
+            } = constraint
+            {
+                if *is_primary {
+                    ensure!(
+                        primary_key_columns.is_none(),
+                        DuplicatePrimaryKey {
+                            first: primary_key_columns.unwrap(),
+                            second: columns.clone()
+                        }
+                    );
+                    primary_key_columns = Some(columns.clone());
+                }
+            }
+        }
+
+        Ok(primary_key_columns)
+    }
+
+    // Find the timestamp column and ensure its valid existence (only one).
+    fn find_and_ensure_timestamp_column(
+        columns_by_name: &BTreeMap<&str, ColumnSchema>,
+        constraints: &[TableConstraint],
+    ) -> Result<Ident> {
+        let mut timestamp_column_name = None;
+        for constraint in constraints {
+            if let TableConstraint::Unique { columns, .. } = constraint {
+                if parser::is_timestamp_key_constraint(constraint) {
+                    // Only one timestamp key constraint
+                    ensure!(timestamp_column_name.is_none(), InvalidTimestampKey);
+                    // Only one column in constraint
+                    ensure!(columns.len() == 1, InvalidTimestampKey);
+                    let timestamp_ident = columns[0].clone();
+
+                    let timestamp_column = columns_by_name
+                        .get(timestamp_ident.value.as_str())
+                        .context(TimestampColumnNotFound {
+                            name: &timestamp_ident.value,
+                        })?;
+
+                    // Ensure the timestamp key's type is timestamp.
+                    ensure!(
+                        timestamp_column.data_type == DatumKind::Timestamp,
+                        InvalidTimestampKey
+                    );
+                    // Ensure the timestamp key is not a tag.
+                    ensure!(
+                        !timestamp_column.is_tag,
+                        TimestampKeyTag {
+                            name: &timestamp_ident.value,
+                        }
+                    );
+
+                    timestamp_column_name = Some(timestamp_ident);
+                }
+            }
+        }
+
+        timestamp_column_name.context(RequireTimestamp)
+    }
+
+    fn create_table_to_plan(&self, stmt: CreateTable) -> Result<Plan> {
+        ensure!(!stmt.table_name.is_empty(), CreateTableNameEmpty);
 
-        let mut schema_builder =
-            schema::Builder::with_capacity(stmt.columns.len()).auto_increment_column_id(true);
+        debug!("Create table to plan, stmt:{:?}", stmt);
 
         // Build all column schemas.
-        let mut name_column_map = stmt
+        let mut columns_by_name = stmt
             .columns
             .iter()
             .map(|col| Ok((col.name.value.as_str(), parse_column(col)?)))
             .collect::<Result<BTreeMap<_, _>>>()?;
```

Comment:
Both `columns_by_name` and `column_idxs_by_name` require no ordering, HashMap is enough.